### PR TITLE
Biz 104 reset password

### PR DESCRIPTION
**Pull request by RomanFedytskyi**

### CODE REVIEW

File: app/components/reset-password-page.js [hunk 1]
```diff
@@ -0,0 +1,25 @@
+import Ember from 'ember';
+
+export default Ember.Component.extend({
+  store: Ember.inject.service(),
```
Comment (dmitrij-borchuk) [inline app/components/reset-password-page.js]: What for?

File: app/components/new-password-page.js [hunk 2]
```diff
@@ -0,0 +1,34 @@
+import Ember from 'ember';
+
+export default Ember.Component.extend({
+  store: Ember.inject.service(),
```
Comment (dmitrij-borchuk) [inline app/components/new-password-page.js]: What for?

File: backend/routing/users.js [hunk 3]
```diff
@@ -30,4 +32,60 @@ const usersController = require('../controllers/users.js')(DAL);
       }
     }
   });
+
+  server.route({
+    method: 'POST',
+    path: '/reset-password',
+    config: {
+      handler: function (request, reply) {
```
Comment (dmitrij-borchuk) [inline backend/routing/users.js:40]: Move to controller

File: backend/routing/users.js [hunk 4]
```diff
@@ -30,4 +32,60 @@ const usersController = require('../controllers/users.js')(DAL);
...
+        let resetToken = utils.newToken();
+        DAL.users.addResetToken(resetToken, request.payload.email).then((response) => {
+          const message = [
+            'Link for reset password: ' + 'http://localhost:4200/new-password/' + resetToken,
+          ].join('\n');
+
+          const mail = {
+            from: '<bizkonect.project@gmail.com>', // sender address
+            to: '<bizkonect.project@gmail.com>', //for test
```
Comment (dmitrij-borchuk) [inline backend/routing/users.js]: How user can receive email?

**Review by dmitrij-borchuk — COMMENTED** [review]
Error: SyntaxError: Unexpected token M in JSON at position 0

After I have clicked send button

File: app/templates/components/login-page.hbs [hunk 5]
```diff
@@ -6,6 +6,7 @@
       {{mdl-textfield label='Password' type="password" value=password}}
       {{#if errorMessage}}
         <p class="error">Login failed: {{errorMessage}}</p>
+        {{#link-to "reset-password"}}reset password{{/link-to}}
```
Comment (dmitrij-borchuk) [inline app/templates/components/login-page.hbs]: Should be shown permanently

File: backend/dal/users.js [hunk 6]
```diff
@@ -55,6 +55,36 @@ module.exports = (connection) => {
...
+          'WHERE email="' + email + '";'
+        ].join('');
+
+        connection.query(request, (err, response) => {
+          err ? reject(err) : resolve(response);
+        });
+      });
+    },
+
+    newPassword: (resetToken, password) => {
```
Comment (dmitrij-borchuk) [inline backend/dal/users.js:78]: Reset token should be cleared

File: backend/dal/users.js [hunk 7]
```diff
@@ -104,5 +134,15 @@ module.exports = (connection) => {
       return connection.query(request, cb);
     },
 
+    addColumn_resetToken: function (cb) {
+      const request = [
+        'ALTER TABLE `users` ',
+        'ADD `resetToken` VARCHAR(255) ',
+        'NOT NULL;'
```
Comment (dmitrij-borchuk) [inline backend/dal/users.js]: Why "not null"?

File: backend/routing/users.js [hunk 8]
```diff
@@ -30,4 +32,60 @@ const usersController = require('../controllers/users.js')(DAL);
...
+
+          Mailer(config.mail).send(mail).then(
+            (res) => {
+              reply(res);
+            }, (err) => {
+              reply( Boom.badImplementation(err.message, err) );
+            }
+          );
+        }, (err) => {
+          reply(err);
```
Comment (dmitrij-borchuk) [inline backend/routing/users.js]: Add Boom

File: backend/routing/users.js [hunk 9]
```diff
@@ -30,4 +32,60 @@ const usersController = require('../controllers/users.js')(DAL);
...
+      handler: function (request, reply) {
+        let resetToken = request.payload.resetToken;
+        let newPassword = request.payload.newPassword;
+        let confirmPassword = request.payload.confirmPassword;
+        if (newPassword === confirmPassword) {
+          DAL.users.newPassword(resetToken, newPassword).then(
+            (res) => {
+              reply();
+            }, (err) => {
+              reply(err);
```
Comment (dmitrij-borchuk) [inline backend/routing/users.js]: Add Boom

**Review by dmitrij-borchuk — CHANGES_REQUESTED** [review]

File: backend/routing/users.js (hunk 4 above)
Comment (RomanFedytskyi) [reply backend/routing/users.js]: it's for testing becouse we have only user with user@user.com password and it isn't valid

File: backend/routing/users.js (hunk 3 above)
Comment (RomanFedytskyi) [reply backend/routing/users.js:40]: fixed

File: backend/dal/users.js (hunk 6 above)
Comment (RomanFedytskyi) [reply backend/dal/users.js:78]: fixed

File: backend/dal/users.js [hunk 10]
```diff
@@ -55,6 +55,44 @@ module.exports = (connection) => {
...
+      return new Promise((resolve, reject) => {
+
+        password = passwordHash.generate(password);
+        let request1 = [
+          'UPDATE `users` ',
+          'SET password="' + password + '" ',
+          'WHERE resetToken="' + resetToken + '";'
+        ].join('');
+
+        let request2 = [
```
Comment (dmitrij-borchuk) [inline backend/dal/users.js]: It can be done in one request

File: backend/controllers/users.js [hunk 11]
```diff
@@ -10,6 +12,31 @@ module.exports = function (DAL) {
...
+      };
+
+      Mailer(config.mail).send(mail).then(
+        (res) => {
+          result = ({"status": "success"});
+        }, (err) => {
+          result = err;
+        }
+      );
+      return result;
```
Comment (dmitrij-borchuk) [inline backend/controllers/users.js]: 'result' will be returned before will be assigned

**Review by dmitrij-borchuk — CHANGES_REQUESTED** [review]

File: backend/controllers/users.js (hunk 11 above)
Comment (dmitrij-borchuk) [reply backend/controllers/users.js]: You should use promice

File: backend/routing/users.js (hunk 3 above)
Comment (dmitrij-borchuk) [reply backend/routing/users.js:40]: Move to controller all logic from handler

**Review by dmitrij-borchuk — CHANGES_REQUESTED** [review]

File: app/templates/components/login-page.hbs [hunk 12]
```diff
@@ -8,6 +8,7 @@
         <p class="error">Login failed: {{errorMessage}}</p>
       {{/if}}
       <div class="button-container">
+        {{#link-to "reset-password"}}Reset{{/link-to}}
```
Comment (dmitrij-borchuk) [inline app/templates/components/login-page.hbs]: Change to "Forgot password"

**Review by dmitrij-borchuk — CHANGES_REQUESTED** [review]

**Review by dmitrij-borchuk — APPROVED** [review]

**Pull request merged**